hs-bindgen/src/HsBindgen/Hs/AST/Name.hs

```diff
@@ -29,50 +62,564 @@ data Namespace =
+-- specially.  Any single quotes in the input are treated as invalid.
+camelCaseCName :: (Char -> String) -> CName -> Text
+camelCaseCName f = T.pack . aux False . T.unpack . getCName
+  where
+    aux :: Bool -> String -> String
+    aux isUp (c:cs)
+      | c == '_'      = aux True cs
+      | isValidChar c = (if isUp then Char.toUpper c else c) : aux False cs
+      | otherwise     = f c ++ aux isUp cs
+    aux isUp []       = if isUp then "_" else "" -- preserve trailing underscore
+
+-- | Drop invalid characters
+dropInvalidChar :: Char -> String
+dropInvalidChar = const ""
+
+-- | Escape invalid characters
+--
+-- An invalid character transformed to a single quote (@'@) followed by the
+-- Unicode code point (four lowercase hex digits).
+escapeInvalidChar :: Char -> String
+escapeInvalidChar c =
+    let hex = showHex (Char.ord c) ""
+    in  '\'' : replicate (max 0 (4 - length hex)) '0' ++ hex
+
+-- | Valid character predicate
+--
+-- Since 'escapeInvalidChar' uses the single quote (@'@) as an escape
+-- character, it is treated specially.  It should never occur in a 'CName',
+-- and this predicate does not consider it valid.
+isValidChar :: Char -> Bool
+isValidChar c = Char.isAlphaNum c || c == '_'
+
+-- | Join parts of a name by concatenating them
+joinWithConcat :: [Text] -> Text
+joinWithConcat = T.concat
+
+-- | Join parts of a name with underscores (@_@)
+joinWithSnakeCase :: [Text] -> Text
+joinWithSnakeCase = T.intercalate "_"
+
+-- | Join parts of a name in @camelCase@ style
+--
+-- The first character of all parts but the first is changed to uppercase (if it
+-- is a letter), and the results are concatenated.
+--
+-- Since this function may change the case of letters, it can cause name
+-- collisions when different C names only differ by case of the first letter.
+joinWithCamelCase :: [Text] -> Text
+joinWithCamelCase = \case
+    (t:ts) -> T.concat $ t : map upperFirstChar ts
+    []     -> T.empty
+  where
+    upperFirstChar :: Text -> Text
+    upperFirstChar t = case T.uncons t of
+      Just (c, t') -> T.cons (Char.toUpper c) t'
+      Nothing      -> t
+
+-- | Do not handle reserved names
+handleReservedNone :: HsName ns -> HsName ns
+handleReservedNone = id
+
+-- | If a name is in a list of reserved names, transform it
+--
+-- The transformation function must return a valid Haskell name.  One
+-- transformation function is provided in this module: 'appendSingleQuote'.
+--
+-- For example, 'keywords' may be used to avoid using Haskell keywords as
+-- identifiers.
+handleReservedNames :: (Text -> Text) -> [Text] -> HsName ns -> HsName ns
+handleReservedNames f reserved name@(HsName t)
+    | t `elem` reserved = HsName $ f t
+    | otherwise         = name
+
+-- | Append a single quote (@'@) to a name
+appendSingleQuote :: Text -> Text
+appendSingleQuote = (<> "'")
+
+-- | Haskell keywords
+--
+-- <https://wiki.haskell.org/Keywords>
+keywords :: [Text]
+keywords =
+    [ "as"
+    , "case"
```

Comment:
cases

Reply:
Thank you very much for the review!

According to the [proposal][]:

> Unlike `case`, `cases` is not a keyword.  Only the token sequence `\` `cases` is treated specially (more precisely, lexically, `cases` is a *varid* rather than a *reservedid*, and it may be used as a variable).

[proposal]: <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0302-cases.rst>

A test indicates that we can indeed uses `cases` as a variable or type variable.

<details><summary>Code</summary>

```haskell
#!/usr/bin/env cabal
{- cabal:
build-depends:
    base
-}
{- project:
with-compiler: ghc-9.6.6
-}

{-# LANGUAGE ImportQualifiedPost #-}
{-# LANGUAGE LambdaCase #-}

module Main (main) where

-- https://hackage.haskell.org/package/base
import Data.Char qualified as Char

------------------------------------------------------------------------------

newtype Foo cases = Foo { unFoo :: String }

cases :: Foo ()
cases = Foo "hHeElLlLoO"

------------------------------------------------------------------------------

filter' :: (a -> Bool) -> [a] -> [a]
filter' = \cases
    _ []          -> []
    p (x:xs)
      | p x       -> x : filter' p xs
      | otherwise ->     filter' p xs

------------------------------------------------------------------------------

main :: IO ()
main = putStrLn $ filter' Char.isLower (unFoo cases)
```

</details>

Perhaps there is a different extension that causes trouble?